File: Group-Project/python/commands.py
"""Prompting, validation, error messages, and output formatting.

One function per command. Each prompts for its fields, rejects bad input with the message the
specification fixes, and asks the same question again rather than abandoning the command.
"""

from typing import Callable, List, Optional, Tuple
import model
import store

# The wording of every rejection. The specification fixes it, and the C++ half prints the same.
_PARSE_MESSAGES = {
    model.DATE_FORM: "Date must be in YYYY-MM-DD form.",
    model.DATE_NOT_REAL: "That date does not exist.",
    model.AMOUNT_FORM: "Amount must be a number.",
    model.AMOUNT_NOT_POSITIVE: "Amount must be greater than zero.",
    model.UNKNOWN_CATEGORY: "Category must be one of: " + ", ".join(model.CATEGORIES) + ".",
}

# What a prompt came back with, when a blank answer and the end of the input both have to be told
# apart from a value.
GIVEN = "given"
BLANK = "blank"
END_OF_INPUT = "end_of_input"


def read_line(prompt: str) -> Optional[str]:
    """Print the prompt and read one trimmed line. None at the end of the input.

    The end of the input abandons whatever command was being typed and leaves the data untouched.
    """
    try:
        return input(prompt).strip()
    except (EOFError, KeyboardInterrupt):
        print()
        return None


def prompt_until_valid(
    prompt: str,
    parse: Callable[[str], object],
    blank_allowed: bool,
) -> Tuple[str, str, object]:
    """Ask until the answer parses or, when blank is allowed, until it is blank.

    Returns what the prompt came back with, the text as it was typed, and the parsed value.
    """
    while True:
        line = read_line(prompt)
        if line is None:
            return END_OF_INPUT, "", None
        if not line and blank_allowed:
            return BLANK, "", None
        try:
            return GIVEN, line, parse(line)
        # ParseError is a ValueError, so this one clause covers every rejection a parser can raise.
        except model.ParseError as error:
            print(_PARSE_MESSAGES[error.reason])


def find_expense(id_text: str) -> Optional[dict]:
    """Look up an expense by its typed id, printing the not-found message when there is none."""
    try:
        expense_id = int(id_text)
    except ValueError:
        print(f"No expense with id {id_text}.")
        return None

    expense = store.get_expense(expense_id)
    if expense is None:
        print(f"No expense with id {id_text}.")
    return expense


def format_table(expense_list: List[dict]) -> str:
    """Format a list of expense dicts into the standard table output."""
    header = f"{'ID':>4}  {'DATE':<10}  {'AMOUNT':>9}  {'CATEGORY':<15}  {'DESCRIPTION'}"
    rows = [header]
    for exp in expense_list:
        amount_str = f"${exp['amount']:.2f}"
        row = (
            f"{exp['id']:>4}  "
            f"{str(exp['date']):<10}  "
            f"{amount_str:>9}  "
            f"{exp['category']:<15}  "
            f"{exp['description']}"
        )
        # An empty description would otherwise leave the padding of the category column dangling.
        rows.append(row.rstrip())
    return "\n".join(rows)


def format_summary(category_totals: List[tuple], overall: float) -> str:
    """Format category totals and overall total into the summary block."""
    lines = []
    for category, total in category_totals:
        amount_str = f"${total:.2f}"
        lines.append(f"{category:<15}{amount_str:>10}")
    lines.append("")
    overall_str = f"${overall:.2f}"
    lines.append(f"{'Overall':<15}{overall_str:>10}")
    return "\n".join(lines)


def add() -> None:
    """Prompt for the four fields in order, validating each, and record a new expense."""
    answer, date_text, _ = prompt_until_valid("Date (YYYY-MM-DD): ", model.parse_date, False)
    if answer == END_OF_INPUT:
        return

    answer, amount_text, _ = prompt_until_valid("Amount: ", model.parse_amount, False)
    if answer == END_OF_INPUT:
        return

    category_prompt = f"Category ({', '.join(model.CATEGORIES)}): "
    answer, category_text, _ = prompt_until_valid(category_prompt, model.parse_category, False)
    if answer == END_OF_INPUT:
        return

    description = read_line("Description: ")
    if description is None:
        return

    # The record is built in this one call, from the four strings exactly as they were typed.
    record = model.parse_expense_fields(date_text, amount_text, category_text, description)
    print(f"Added expense {store.add_expense(record)}.")


def list_expenses() -> None:
    """Print every expense sorted by id ascending."""
    items = store.list_expenses()
    if not items:
        print("No expenses.")
    else:
        print(format_table(items))


def filter_expenses() -> None:
    """Prompt for a date range and a category, any of which may be blank, and print the matches."""
    answer, _, start_date = prompt_until_valid(
        "Start date (YYYY-MM-DD, blank for none): ", model.parse_date, True
    )
    if answer == END_OF_INPUT:
        return

    answer, _, end_date = prompt_until_valid(
        "End date (YYYY-MM-DD, blank for none): ", model.parse_date, True
    )
    if answer == END_OF_INPUT:
        return

    answer, _, category = prompt_until_valid(
        "Category (blank for all): ", model.parse_category, True
    )
    if answer == END_OF_INPUT:
        return

    matches = store.filter_expenses(start_date=start_date, end_date=end_date, category=category)
    if not matches:
        print("No expenses match.")
    else:
        print(format_table(matches))


def summary() -> None:
    """Print a total for each category and an overall total."""
    category_totals, overall = store.get_summary()
    print(format_summary(category_totals, overall))


def edit() -> None:
    """Prompt for an id and each editable field, where a blank answer keeps the stored value."""
    id_text = read_line("Expense id: ")
    if id_text is None:
        return

    expense = find_expense(id_text)
    if expense is None:
        return

    answer, _, date = prompt_until_valid(
        f"Date [{expense['date']}] (blank to keep): ", model.parse_date, True
    )
    if answer == END_OF_INPUT:
        return
    if answer == BLANK:
        date = expense["date"]

    answer, _, amount = prompt_until_valid(
        f"Amount [{expense['amount']:.2f}] (blank to keep): ", model.parse_amount, True
    )
    if answer == END_OF_INPUT:
        return
    if answer == BLANK:
        amount = expense["amount"]

    answer, _, category = prompt_until_valid(
        f"Category [{expense['category']}] (blank to keep): ", model.parse_category, True
    )
    if answer == END_OF_INPUT:
        return
    if answer == BLANK:
        category = expense["category"]

    description = read_line(f"Description [{expense['description']}] (blank to keep): ")
    if description is None:
        return
    if not description:
        description = expense["description"]

    # A kept answer carries the stored value straight over, so keeping an amount cannot round it.
    updated = {
        "date": date,
        "amount": amount,
        "category": category,
        "description": description,
    }
    expense_id = expense["id"]
    store.update_expense(expense_id, updated)
    print(f"Updated expense {expense_id}.")


def delete_expense() -> None:
    """Prompt for an id and remove that expense."""
    id_text = read_line("Expense id: ")
    if id_text is None:
        return

    expense = find_expense(id_text)
    if expense is None:
        return

    expense_id = expense["id"]
    store.delete_expense(expense_id)
    print(f"Deleted expense {expense_id}.")

File: Group-Project/cpp/commands.cpp
#include "commands.h"

#include <functional>
#include <iomanip>
#include <iostream>
#include <optional>
#include <sstream>
#include <stdexcept>

namespace commands {
namespace {

const char* const kAmountNotNumber = "Amount must be a number.";
const char* const kAmountNotPositive = "Amount must be greater than zero.";
const char* const kDateForm = "Date must be in YYYY-MM-DD form.";
const char* const kDateNotReal = "That date does not exist.";

std::string trim(const std::string& text) {
    const std::size_t first = text.find_first_not_of(" \t\r\n");
    if (first == std::string::npos) {
        return "";
    }
    const std::size_t last = text.find_last_not_of(" \t\r\n");
    return text.substr(first, last - first + 1);
}

// Prints the prompt and reads one trimmed line. False at the end of the input, which abandons
// whatever command was being typed and leaves the data untouched.
bool read_line(const std::string& prompt, std::string& line) {
    std::cout << prompt << std::flush;
    std::string raw;
    if (!std::getline(std::cin, raw)) {
        std::cout << "\n";
        return false;
    }
    line = trim(raw);
    return true;
}

std::string category_error() {
    return "Category must be one of: " + category_names_joined() + ".";
}

std::string not_found_error(const std::string& id_text) {
    return "No expense with id " + id_text + ".";
}

void print_parse_error(ParseError error) {
    switch (error) {
        case ParseError::DateForm:          std::cout << kDateForm << "\n"; break;
        case ParseError::DateNotReal:       std::cout << kDateNotReal << "\n"; break;
        case ParseError::AmountForm:        std::cout << kAmountNotNumber << "\n"; break;
        case ParseError::AmountNotPositive: std::cout << kAmountNotPositive << "\n"; break;
        case ParseError::UnknownCategory:   std::cout << category_error() << "\n"; break;
        case ParseError::None:              break;
    }
}

// What a prompt that can be left blank came back with.
enum class Answer {
    Given,
    Blank,
    EndOfInput
};

// Asks until the answer parses or, when blank is allowed, until it is blank. The parser writes
// the value where its caller wants it; all this loop needs back is whether the text was accepted.
Answer prompt_until_valid(const std::string& prompt, bool blank_allowed,
                          const std::function<ParseError(const std::string&)>& parse) {
    std::string line;
    while (read_line(prompt, line)) {
        if (line.empty() && blank_allowed) {
            return Answer::Blank;
        }
        const ParseError error = parse(line);
        if (error == ParseError::None) {
            return Answer::Given;
        }
        print_parse_error(error);
    }
    return Answer::EndOfInput;
}

std::string table_header() {
    std::ostringstream row;
    row << std::right << std::setw(4) << "ID" << "  "
        << std::left << std::setw(10) << "DATE" << "  "
        << std::right << std::setw(9) << "AMOUNT" << "  "
        << std::left << std::setw(15) << "CATEGORY" << "  "
        << "DESCRIPTION";
    return row.str();
}

std::string table_row(const Expense& expense) {
    std::ostringstream row;
    row << std::right << std::setw(4) << expense.id << "  "
        << std::left << std::setw(10) << format_date(expense.date) << "  "
        << std::right << std::setw(9) << format_amount(expense.amount) << "  "
        << std::left << std::setw(15) << category_name(expense.category) << "  "
        << expense.description;
    // An empty description would otherwise leave the padding of the category column dangling.
    std::string line = row.str();
    return line.substr(0, line.find_last_not_of(' ') + 1);
}

void print_table(const std::vector<const Expense*>& expenses) {
    std::cout << table_header() << "\n";
    for (const Expense* expense : expenses) {
        std::cout << table_row(*expense) << "\n";
    }
}

// The whole of the text as an integer, or nothing when it is not one. std::stoi stops at the
// first character it cannot use, so a partial read counts as no read.
std::optional<int> read_int(const std::string& text) {
    try {
        std::size_t consumed = 0;
        const int value = std::stoi(text, &consumed);
        if (consumed == text.size()) {
            return value;
        }
    } catch (const std::invalid_argument&) {
        // falls through to nothing
    } catch (const std::out_of_range&) {
        // falls through to nothing
    }
    return std::nullopt;
}

// Asks for an id and hands back the expense carrying it, printing the not-found message itself and
// quoting the text as it was typed. Null at the end of the input too, since both leave the command.
const Expense* prompt_for_expense(const ExpenseStore& store) {
    std::string id_text;
    if (!read_line("Expense id: ", id_text)) {
        return nullptr;
    }
    const std::optional<int> id = read_int(id_text);
    const Expense* found = id.has_value() ? store.find(*id) : nullptr;
    if (found == nullptr) {
        std::cout << not_found_error(id_text) << "\n";
    }
    return found;
}

}  // namespace

void print_menu() {
    std::cout << "Expense Tracker\n"
              << "Commands: add, list, filter, summary, edit, delete, quit\n";
}

bool read_command(std::string& command, std::string& command_as_typed) {
    if (!read_line("> ", command_as_typed)) {
        return false;
    }
    command = to_lower(command_as_typed);
    return true;
}

void add(ExpenseStore& store) {
    Date date{};
    if (prompt_until_valid("Date (YYYY-MM-DD): ", false,
                           [&date](const std::string& line) { return parse_date(line, date); })
        != Answer::Given) {
        return;
    }
    double amount = 0.0;
    if (prompt_until_valid("Amount: ", false,
                           [&amount](const std::string& line) { return parse_amount(line, amount); })
        != Answer::Given) {
        return;
    }
    Category category = Category::Other;
    const std::string category_prompt = "Category (" + category_names_joined() + "): ";
    if (prompt_until_valid(category_prompt, false,
                           [&category](const std::string& line) {
                               return parse_category(line, category);
                           })
        != Answer::Given) {
        return;
    }
    std::string description;
    if (!read_line("Description: ", description)) {
        return;
    }

    // Nothing reaches the store until every field has passed.
    const int id = store.add(date, amount, category, description);
    std::cout << "Added expense " << id << ".\n";
}

void list_expenses(const ExpenseStore& store) {
    // No criterion at all, which is what the specification says a wholly blank filter comes to.
    const std::vector<const Expense*> expenses = store.filter(FilterCriteria{});
    if (expenses.empty()) {
        std::cout << "No expenses.\n";
        return;
    }
    print_table(expenses);
}

void filter_expenses(const ExpenseStore& store) {
    FilterCriteria criteria;

    Date start{};
    const Answer start_answer =
        prompt_until_valid("Start date (YYYY-MM-DD, blank for none): ", true,
                           [&start](const std::string& line) { return parse_date(line, start); });
    if (start_answer == Answer::EndOfInput) {
        return;
    }
    if (start_answer == Answer::Given) {
        criteria.start = start;
    }

    Date end{};
    const Answer end_answer =
        prompt_until_valid("End date (YYYY-MM-DD, blank for none): ", true,
                           [&end](const std::string& line) { return parse_date(line, end); });
    if (end_answer == Answer::EndOfInput) {
        return;
    }
    if (end_answer == Answer::Given) {
        criteria.end = end;
    }

    Category category = Category::Other;
    const Answer category_answer =
        prompt_until_valid("Category (blank for all): ", true,
                           [&category](const std::string& line) {
                               return parse_category(line, category);
                           });
    if (category_answer == Answer::EndOfInput) {
        return;
    }
    if (category_answer == Answer::Given) {
        criteria.category = category;
    }

    const std::vector<const Expense*> matches = store.filter(criteria);
    if (matches.empty()) {
        std::cout << "No expenses match.\n";
        return;
    }
    print_table(matches);
}

void summary(const ExpenseStore& store) {
    for (Category category : all_categories()) {
        std::cout << std::left << std::setw(15) << category_name(category)
                  << std::right << std::setw(10) << format_amount(store.category_total(category))
                  << "\n";
    }
    std::cout << "\n"
              << std::left << std::setw(15) << "Overall"
              << std::right << std::setw(10) << format_amount(store.overall_total()) << "\n";
}

void edit(ExpenseStore& store) {
    const Expense* current = prompt_for_expense(store);
    if (current == nullptr) {
        return;
    }

    // A copy, so an abandoned command leaves the stored expense as it was.
    Expense edited = *current;

    Date date{};
    const Answer date_answer =
        prompt_until_valid("Date [" + format_date(edited.date) + "] (blank to keep): ", true,
                           [&date](const std::string& line) { return parse_date(line, date); });
    if (date_answer == Answer::EndOfInput) {
        return;
    }
    if (date_answer == Answer::Given) {
        edited.date = date;
    }

    double amount = 0.0;
    const Answer amount_answer =
        prompt_until_valid("Amount [" + format_amount_plain(edited.amount) + "] (blank to keep): ",
                           true,
                           [&amount](const std::string& line) { return parse_amount(line, amount); });
    if (amount_answer == Answer::EndOfInput) {
        return;
    }
    if (amount_answer == Answer::Given) {
        edited.amount = amount;
    }

    Category category = Category::Other;
    const Answer category_answer =
        prompt_until_valid("Category [" + category_name(edited.category) + "] (blank to keep): ",
                           true,
                           [&category](const std::string& line) {
                               return parse_category(line, category);
                           });
    if (category_answer == Answer::EndOfInput) {
        return;
    }
    if (category_answer == Answer::Given) {
        edited.category = category;
    }

    std::string description;
    if (!read_line("Description [" + edited.description + "] (blank to keep): ", description)) {
        return;
    }
    if (!description.empty()) {
        edited.description = description;
    }

    store.update(edited);
    std::cout << "Updated expense " << edited.id << ".\n";
}

void delete_expense(ExpenseStore& store) {
    const Expense* target = prompt_for_expense(store);
    if (target == nullptr) {
        return;
    }
    // The id came out of the store, so the removal cannot miss, and the pointer dies with it.
    const int id = target->id;
    store.delete_expense(id);
    std::cout << "Deleted expense " << id << ".\n";
}

void unknown(const std::string& command) {
    std::cout << "Unknown command: " << command << ".\n";
    print_menu();
}

}  // namespace commands
